frida_tools/tracer.py

```diff
@@ -678,7 +700,7 @@ def _create_trace_script(self):
         return {};
     }
 }
-"""
+""" % {"userdata": self._userdata}
```

Comment:
```suggestion
"""
```

So instead of passing it in here, we should call the `init()` method right after we have done `self._script.load()`:

```py
        self._script.exports.init(stage, parameters)
```

So we'll have to pass in `stage` and `parameters` to `start_trace()`:

```py
    def start_trace(self, session, runtime, stage, parameters, ui):
```

And the caller will need to compute `stage`, like we do in the REPL:

```py
        stage = 'early' if self._target[0] == 'file' and is_first_load else 'late'
```

(We should push this down into the `ConsoleApplication` base class.)